Код ревью приложения

SeriousMeow opened the pull request.

SeriousMeow committed:
Инициализируй проект

SeriousMeow committed:
Добавь библиотеку Renderer

SeriousMeow committed:
Добавь библиотеку Qt6

SeriousMeow committed:
Добавь сборку с инструментами Qt6

SeriousMeow committed:
Создай структуру приложения

SeriousMeow committed:
Обнови версию библиотеки Renderer

SeriousMeow committed:
Добавь порты

SeriousMeow committed:
Добавь связь между сценой и областью отрисовки

DimaTrushin commented on source/main.cpp:
```diff
@@ -0,0 +1,15 @@
+
+int main(int argc, char* argv[]) {
+    try {
+        QApplication app(argc, argv);
```
Я бы вынес этот объект из try блока. Это связано с тем, что ты в catch не сомжешь использовать его ресурсы, например. И назови переменную qruntime, а свое приложение app.

DimaTrushin commented on source/interface/view.hpp:
```diff
@@ -0,0 +1,63 @@
+    QLabel* render_region_;
+    ports::PortIn<renderer::Scene> scene_port_;
+    renderer::Scene::CameraId camera_id_ = 0;
+    renderer::Renderer renderer_;
```
Какого хрена у тебя рендерер лежит во view? Мало того, что view должна быть stateless, так ты еще и часть ядра сюда поместил. Убрать.

DimaTrushin commented on source/interface/view.hpp:
```diff
@@ -0,0 +1,63 @@
+    void RenderScene(const renderer::Scene& scene);
+
+    QLabel* render_region_;
+    ports::PortIn<renderer::Scene> scene_port_;
```
Не используй длинные мучительные имена для доступа к типам. Вводи в начале класса в приватной части все юзинги необходимые для работы класса, например.
```cpp
using Scene = renderer::Scene;
using ScenePort = ports::PortIn<Scene>;
...
```
Тогда это повысит читаемость кода.

DimaTrushin commented on source/interface/view.hpp:
```diff
@@ -0,0 +1,63 @@
+/**
+ * @brief Класс взаимодействия с областью отрисовки
+ */
+class View {
```
Судя по тому что я вижу, это НЕ View, это что-то странное.

DimaTrushin commented on source/interface/view.cpp:
```diff
@@ -0,0 +1,54 @@
+
+View::View(QLabel* render_region)
+    : render_region_{render_region},
+      scene_port_{[this](const renderer::Scene& scene) { RenderScene(scene); },
```
Если тебе надо выполнять одно действие, то сделай специальный конструктор, который принимает одно действие и записывает его во все колбэки.

DimaTrushin commented on source/interface/view.cpp:
```diff
@@ -0,0 +1,54 @@
+        assert((image.GetHeight() == render_region_->size().height()) and
+               "Draw: высота изображения не совпадает с высотой области отрисовки");
+    }
+    QImage intermediate_image{static_cast<int>(image.GetWidth()),
```
Касты в коде -- это обычно признак ошибки дизайна, твои классы сообщают данные не в том формате, в котором их используют другие объекты.

DimaTrushin replied:
У тебя куча ненужной левой деятельности происходит. Я не понимаю зачем тут все эти кишки нужны. Я бы видел это все в виде двух строчек
```cpp
QPixmap picture = convertToQPixmap(image);
render_region_->setPixmap(picture);
```
И не надо тут размазывать кишки конвертации. Далее
```cpp
QPixmap convertToQPixmap(const Image& image) const {
  QImage qimage = convertToQImage(image); // и вот тут ты делаешь всю грязную работу
  return QPixelmap::fromImage(qimage); // это статическая функция, тут не надо вызывать на объекте класса
}
```

DimaTrushin commented on source/interface/view.cpp:
```diff
@@ -0,0 +1,54 @@
+    // Изображения имеют одинаковый размер, нужно сконвертировать пиксели одного в другое
+    // Чтобы избавиться от проверок со стороны Qt, запись идет сразу во внутреннюю память
+    // Для Format_RGB32 пиксели хранятся строками сверху вниз
+    QRgb* intermediate_data = reinterpret_cast<QRgb*>(intermediate_image.bits());
```
А это точно не UB?

DimaTrushin commented on source/interface/view.hpp:
```diff
@@ -0,0 +1,63 @@
+     */
+    void RenderScene(const renderer::Scene& scene);
+
+    QLabel* render_region_;
```
Кто владеет этими данными?

DimaTrushin commented on source/interface/view.cpp:
```diff
@@ -0,0 +1,54 @@
+    return &scene_port_;
+}
+
+void View::RenderScene(const renderer::Scene& scene) {
```
Этому не место во View, это часть ядра.

DimaTrushin commented on source/interface/view.cpp:
```diff
@@ -0,0 +1,54 @@
+    QSize region_size = render_region_->size();
+    renderer::Image image{renderer::Width{static_cast<size_t>(region_size.width())},
+                          renderer::Height{static_cast<size_t>(region_size.height())}};
+    if (scene.HasCamera(camera_id_)) {
```
Если пользователь дал тебе отрендерить некорректную камеру, это ошибка программиста. Это надо ассертить, а не if-ать.

DimaTrushin replied:
Плохо сохранять camera_id между вызовами. Потому что у тебя сцена могла поменяться и id для камеры ссылается на камеру в первой сцене, а у тебя передали вторую сцену. И все, у тебя сломался id и он указывает на несуществующую камеру. Так лучше не делать.

Если ты хочешь сделать активную камеру, то у тебя должна быть и привязка к текущей сцене. Чтобы эту зависимость сделать надежной, можно сделать так. У тебя есть источник сцены SceneModel и он через порт оповещает RenderModel о состоянии текущей сцены через observable/observer порт. При этом RenderModel хранит active_camera_id. Но если от SceneModel прилетает другая сцена, то active_camera_id сбрасывается в некорректное состояние (например в -1).

Либо если ты не хочешь вводить такие мелкие модели, то тебе надо хранить дополнительно scene_id, чтобы можно было понять изменилась сцена или нет. И если изменилась, то ты сбрасываешь active_camera_id.

DimaTrushin commented on source/interface/view.cpp:
```diff
@@ -0,0 +1,54 @@
+    renderer::Image image{renderer::Width{static_cast<size_t>(region_size.width())},
+                          renderer::Height{static_cast<size_t>(region_size.height())}};
+    if (scene.HasCamera(camera_id_)) {
+        image = renderer_.Render(scene, camera_id_, std::move(image));
```
это должно делаться в ядре

DimaTrushin commented on source/interface/view.cpp:
```diff
@@ -0,0 +1,54 @@
+    const renderer::Image::Pixel* image_data = image.AccessData();
+    for (size_t i = 0; i < image.GetHeight() * image.GetWidth(); ++i) {
+        intermediate_data[i] = qRgb(image_data[i].r, image_data[i].g, image_data[i].b);
+    }
```
Что тут за хуйня происходит? Тут явно конвертируются данные. Что во что конвертируется?

DimaTrushin commented on source/except.cpp:
```diff
@@ -0,0 +1,15 @@
+    } catch (...) {
+        std::cerr << "Unknown error" << '\n';
+    }
+    return 1;
```
зачем?

DimaTrushin commented on source/app_kernel.hpp:
```diff
@@ -0,0 +1,22 @@
+    AppKernel();
+
+protected:
+    app::models::Scene scene_;
```
Зачем тебе app:: когда ты и так находишься в app namespace?

DimaTrushin commented on source/app_gui.hpp:
```diff
@@ -0,0 +1,24 @@
+
+protected:
+    MainWindow main_window_;
+    View view_;
```
А View это не часть основного окна?

DimaTrushin commented on source/models/3D_scene.hpp:
```diff
@@ -0,0 +1,33 @@
+#include "../utils/ports.hpp"
+
+namespace app {
+namespace models {
```
Зачем тебе выделять модели в отдельный namespace?

DimaTrushin commented on source/models/3D_scene.cpp:
```diff
@@ -0,0 +1,12 @@
+
+namespace app::models {
+
+void Scene::SubscribeScenePort(ports::PortIn<renderer::Scene>* port) {
```
можно убрать слово port из названия метода, потому что и так понятно, что ты к порту подписываешься.

DimaTrushin reviewed (changes requested):
Обрати внимание, что я не повторял замечания. Если одна и та же ошибка всречалась несколько раз, я комментировал лишь единожды. Потому я ожидаю, что ты поправишь все аналогичные замечания всюду.

SeriousMeow committed:
Добавь движок для работы со сценой и интерфейс для управления

SeriousMeow committed:
Добавь функциональность портам

Теперь через подписанный PortIn можно получить лежащие в PortOut данные

SeriousMeow committed:
Исправь ошибку монитора камеры

Исправлена проблема, когда при приходе информации об изменении камеры
после изменения сцены информация оставалась о старой камере

SeriousMeow committed:
Добавь интерфейс управления объектом

SeriousMeow committed:
Добавь создание камер и выбор камер и объектов из списка

SeriousMeow committed:
Исправь ошибку добавления камеры

Исправлена ошибка, из-за которой изображение с камеры не обновлялось до
первого изменения

SeriousMeow committed:
Исправь ошибку падения приложения и убери кнопки задания параметров

Исправлена ошибка, когда приложение могло падать при запуске или при
изменении параметров объекта при отсутсвии объектов в сцене

Также удалены кнопки задания параметров. Теперь обновление происходит
автоматически при изменении полей

SeriousMeow committed:
Обнови версию Renderer, добавь статическое освещение и исправь ошибки

SeriousMeow committed:
Исправь ошибку сбоя координат при переключении объектов

SeriousMeow committed:
Удали m.css и исправь генерацию документации

SeriousMeow committed:
Обнови README

SeriousMeow committed:
Добавь новый конструктор порта и упрости код

SeriousMeow committed:
Распредели конвертацию в View::Draw по функциям

SeriousMeow committed:
Переименуй View в ImageRegion

SeriousMeow committed:
Измени структуру main

SeriousMeow committed:
Перенеси на итоговую версию Renderer
